tracing-opentelemetry/src/main/java/io/micronaut/tracing/opentelemetry/interceptor/TraceInterceptor.java

```diff
@@ -0,0 +1,218 @@
+
+import io.micronaut.aop.InterceptPhase;
+import io.micronaut.aop.InterceptedMethod;
+import io.micronaut.aop.InterceptorBean;
+import io.micronaut.aop.MethodInterceptor;
+import io.micronaut.aop.MethodInvocationContext;
+import io.micronaut.context.annotation.Requires;
+import io.micronaut.core.annotation.AnnotationMetadata;
+import io.micronaut.core.annotation.AnnotationValue;
+import io.micronaut.core.annotation.NonNull;
+import io.micronaut.core.annotation.Nullable;
+import io.micronaut.core.type.Argument;
+import io.micronaut.core.util.StringUtils;
+import io.micronaut.tracing.annotation.ContinueSpan;
+import io.micronaut.tracing.annotation.NewSpan;
+import io.micronaut.tracing.annotation.SpanTag;
+import io.micronaut.tracing.opentelemetry.instrument.util.TracingPublisher;
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.api.trace.Tracer;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.instrumentation.api.instrumenter.Instrumenter;
+import io.opentelemetry.instrumentation.api.util.ClassAndMethod;
+import jakarta.inject.Named;
+import jakarta.inject.Singleton;
+import org.reactivestreams.Publisher;
+
+import java.util.concurrent.CompletionStage;
+
+/**
+ * Implements tracing logic for <code>ContinueSpan</code> and <code>NewSpan</code>
+ * using the Open Tracing API.
+ *
+ * @author Nemanja Mikic
+ * @since 4.1.0
+ */
+@Singleton
+@Requires(beans = Tracer.class)
+@InterceptorBean({ContinueSpan.class, NewSpan.class})
+public class TraceInterceptor implements MethodInterceptor<Object, Object> {
+
+    private final Instrumenter<ClassAndMethod, Object> instrumenter;
+
+    /**
+     * Initialize the interceptor with tracer and conversion service.
+     *
+     * @param instrumenter the ClassAndMethod Instrumenter
+     */
+    public TraceInterceptor(@Named("micronautCodeTelemetryInstrumenter") Instrumenter<ClassAndMethod, Object> instrumenter) {
+        this.instrumenter = instrumenter;
+    }
+
+    @Override
+    public int getOrder() {
+        return InterceptPhase.TRACE.getPosition();
+    }
+
+    /**
+     * Logs an error to the span.
+     *
+     * @param context the span
+     * @param e       the error
+     */
+    public static void logError(Context context, Throwable e) {
+        Span.fromContext(context).recordException(e);
+    }
+
+    @Nullable
+    @SuppressWarnings("unchecked")
+    @Override
+    public Object intercept(MethodInvocationContext<Object, Object> context) {
+        boolean isContinue = context.hasAnnotation(ContinueSpan.class);
+        AnnotationValue<NewSpan> newSpan = context.getAnnotation(NewSpan.class);
+        boolean isNew = newSpan != null;
+        if (!isContinue && !isNew) {
+            return context.proceed();
+        }
+        Context currentContext = Context.current();
+
+        ClassAndMethod classAndMethod = ClassAndMethod.create(context.getDeclaringType(), context.getMethodName());
```

Comment:
Just nitpicking but this can be moved copied `PUBLISHER` block and created if needed in `else` block